src/thin_arc.rs

```diff
@@ -234,13 +222,60 @@ impl<H, T> Arc<HeaderSliceWithLength<H, [T]>> {
             a.slice.len(),
             "Length needs to be correct for ThinArc to work"
         );
+        // Safety: invariant checked in assertion above
         unsafe { Self::into_thin_unchecked(a) }
     }
 
     /// Converts a `ThinArc` into an `Arc`. This consumes the `ThinArc`, so the refcount
     /// is not modified.
     #[inline]
     pub fn from_thin(a: ThinArc<H, T>) -> Self {
+        Self::from_checked_header(
+            Arc::<HeaderSliceWithLengthChecked<H, [T]>>::checked_header_from_thin(a),
+        )
+    }
+
+    /// Converts an `Arc` into a `ThinArc`. This consumes the `Arc`, so the refcount
+    /// is not modified.
+    ///
+    /// # Safety
+    /// Assumes that the header length matches the slice length.
+    #[inline]
+    fn from_checked_header(a: Arc<HeaderSliceWithLengthChecked<H, [T]>>) -> Self {
+        // Safety: HeaderSliceWithLengthChecked and HeaderSliceWithLengthUnchecked have the same layout
+        unsafe { mem::transmute(a) }
+    }
+}
+
+impl<H, T> Arc<HeaderSliceWithLengthChecked<H, [T]>> {
+    /// Converts an `Arc` into a `ThinArc`. This consumes the `Arc`, so the refcount
+    /// is not modified.
+    #[inline]
+    pub fn checked_header_into_thin(a: Self) -> ThinArc<H, T> {
+        debug_assert_eq!(
+            a.length(),
+            a.slice().len(),
+            "Length needs to be correct for ThinArc to work"
+        );
+
+        let a = ManuallyDrop::new(a);
+        let fat_ptr: *mut ArcInner<HeaderSliceWithLengthChecked<H, [T]>> = a.ptr();
+        let thin_ptr = fat_ptr as *mut [usize] as *mut usize;
+        ThinArc {
+            // Safety: The pointer comes from a valid Arc, and HeaderSliceWithLengthChecked has the correct length invariant
+            ptr: unsafe {
+                ptr::NonNull::new_unchecked(
+                    thin_ptr as *mut ArcInner<HeaderSliceWithLengthChecked<H, [T; 0]>>,
+                )
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// Converts a `ThinArc` into an `Arc`. This consumes the `ThinArc`, so the refcount
+    /// is not modified.
+    #[inline]
+    pub fn checked_header_from_thin(a: ThinArc<H, T>) -> Self {
```

Comment:
I’m not *really* happy with the naming, haven’t thought of better alternatives yet.

The problem is that now there’s `Arc::from_thin` and `Arc::checked_header_from_thin`, and the latter just *sounds* more heavy, like there’s additional checking going on, even though the exact *opposite* is the case (you get something that needs *less* checking in order to convert back).

---

Actually, now I have an idea: how about the word becomes “protected”?

E.g. `HeaderSliceWithLengthProtected` instead of `HeaderSliceWithLengthChecked`. (And the `HeaderSliceWithLengthUnchecked` type synonym isn’t needed at all.)

Then we can have `Arc::from_thin` and `Arc::protected_from_thin`. That sounds less like additional “checking” is going on.

And we can have `ThinArc::with_arc` and `ThinArc::with_protected_arc` (the latter can also serve to be used internally an implementation that replaces the unsafe check-less `clone` from #76). We could even consider to *keep* `with_arc_mut` [with sufficient checks, i.e. under documented threat of abort] and add `with_protected_arc_mut` next to it, making the naming more consistent.

----

Or (coming back to naming) another idea is `HeaderSliceWithConstLen`, and thus `Arc::const_len_from_thin`, `ThinArc::with_const_len_arc` and `ThinArc::with_const_len_arc_mut`.

Reply:
Yeah I'm not happy with the naming either, this is just the naming that's stable. The unchecked variant should be more verbose.

Reply:
Protected seems fine for now. I don't like the implication of `const` here.